pkg/operator/sapinstancestart_v1.go

```diff
@@ -0,0 +1,290 @@
+	}
+
+	return waitUntilSapInstanceState(
+		ctx,
+		s.sapControlConnector,
+		sapcontrol.STATECOLORSAPControlGRAY,
+		s.parsedArguments.timeout,
+		s.interval,
+	)
+}
+
+func (s *SAPInstanceStart) operationDiff(ctx context.Context) map[string]any {
+	diff := make(map[string]any)
+
+	beforeDiffOutput := sapInstanceStartDiffOutput{
+		Started: s.resources[beforeDiffField].(bool),
+	}
+	before, _ := json.Marshal(beforeDiffOutput)
+	diff["before"] = string(before)
+
+	afterDiffOutput := sapInstanceStartDiffOutput{
+		Started: s.resources[afterDiffField].(bool),
+	}
+	after, _ := json.Marshal(afterDiffOutput)
+	diff["after"] = string(after)
+
+	return diff
+}
+
+func allProcessesInState(
+	ctx context.Context,
+	connector sapcontrol.SAPControlConnector,
+	expectedState sapcontrol.STATECOLOR,
+) (bool, error) {
+	request := new(sapcontrol.GetProcessList)
+	response, err := connector.GetProcessListContext(ctx, request)
+	if err != nil {
+		return false, fmt.Errorf("error getting instance process list: %w", err)
+	}
+
+	for _, process := range response.Processes {
+		if *process.Dispstatus != expectedState {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func waitUntilSapInstanceState(
+	ctx context.Context,
+	connector sapcontrol.SAPControlConnector,
+	expectedState sapcontrol.STATECOLOR,
+	timeout time.Duration,
+	interval time.Duration,
+) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	for {
+		inState, err := allProcessesInState(timeoutCtx, connector, expectedState)
+		if err != nil {
+			return err
+		}
+
+		if timeoutCtx.Err() != nil {
+			return fmt.Errorf("error waiting until instance is in desired state")
+		}
+
+		if inState {
+			return nil
+		}
+
+		err = sleepContext(timeoutCtx, interval)
+		if err != nil {
+			return err
+		}
+	}
+
+}
+
+// sleepContext sleeps the running thread until the interval or the context are completed
+func sleepContext(ctx context.Context, interval time.Duration) error {
+	timer := time.NewTimer(interval)
+	select {
+	case <-ctx.Done():
+		timer.Stop()
+		return ctx.Err()
+	case <-timer.C:
+		return nil
+	}
+}
+
+func parseSAPStateChangeArguments(rawArguments OperatorArguments) (*sapStateChangeArguments, error) {
+	instNumberArgument, found := rawArguments["instance_number"]
+	if !found {
+		return nil, fmt.Errorf("argument instance_number not provided, could not use the operator")
+	}
+
+	instNumber, ok := instNumberArgument.(string)
+	if !ok {
+		return nil, fmt.Errorf(
+			"could not parse instance_number argument as string, argument provided: %v",
+			instNumberArgument,
+		)
+	}
+
+	timeout := defaultSapInstanceStateTimeout
+	if timeoutArgument, found := rawArguments["timeout"]; found {
+		timeoutFloat, ok := timeoutArgument.(float64)
+		if !ok {
+			return nil, fmt.Errorf(
+				"could not parse timeout argument as a number, argument provided: %v",
+				timeoutArgument,
+			)
+		}
+
+		timeout = time.Duration(int(timeoutFloat)) * time.Second
```

Comment:
What makes us need the double casting of the timeout float -> int?

Reply:
Yes, it is not needed. I thought that the Duration casting needed an int, but a float as well works

Reply:
Updated